src/EoS/Nuclear/CPU_EoS_Nuclear.cpp

```diff
@@ -0,0 +1,579 @@
+{
+
+   if ( x != x  ||  x < -__FLT_MAX__  ||  x > __FLT_MAX__ )    return true;
+   else                                                        return false;
+
+} // FUNCTION : Nuc_Overflow
+#endif // #ifdef GAMER_DEBUG
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  EoS_DensEint2Pres_Nuclear
+// Description :  Convert gas mass density and internal energy density to gas pressure
+//
+// Note        :  1. Internal energy density here is per unit volume instead of per unit mass
+//                2. See EoS_SetAuxArray_Nuclear() for the values stored in AuxArray_Flt/Int[]
+//
+// Parameter   :  Dens_Code    : Gas mass density            (in code unit)
+//                Eint_Code    : Gas internal energy density (in code unit)
+//                Passive_Code : Passive scalars             (in code unit)
+//                AuxArray_*   : Auxiliary arrays (see the Note above)
+//                Table        : EoS tables
+//
+// Return      :  Gas pressure (in code unit)
+//-------------------------------------------------------------------------------------------------------
+GPU_DEVICE_NOINLINE
+static real EoS_DensEint2Pres_Nuclear( const real Dens_Code, const real Eint_Code, const real Passive_Code[],
+                                       const double AuxArray_Flt[], const int AuxArray_Int[],
+                                       const real *const Table[EOS_NTABLE_MAX] )
+{
+
+// check
+#  ifdef GAMER_DEBUG
+#  if ( NCOMP_PASSIVE > 0 )
+   if ( Passive_Code == NULL )   printf( "ERROR : Passive_Code == NULL in %s !!\n", __FUNCTION__ );
+#  endif
+   if ( AuxArray_Flt == NULL )   printf( "ERROR : AuxArray_Flt == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Int == NULL )   printf( "ERROR : AuxArray_Int == NULL in %s !!\n", __FUNCTION__ );
+
+   if ( Hydro_CheckNegative(Dens_Code) )
+      printf( "ERROR : invalid input density (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+              Dens_Code, __FILE__, __LINE__, __FUNCTION__ );
+
+// still require Eint>0 for the nuclear EoS
+   if ( Hydro_CheckNegative(Eint_Code) )
+      printf( "ERROR : invalid input internal energy (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+              Eint_Code, __FILE__, __LINE__, __FUNCTION__ );
+#  endif // GAMER_DEBUG
+
+
+   const real EnergyShift = AuxArray_Flt[NUC_AUX_ESHIFT   ];
+   const real Dens2CGS    = AuxArray_Flt[NUC_AUX_DENS2CGS ];
+   const real sEint2CGS   = AuxArray_Flt[NUC_AUX_VSQR2CGS ];
+   const real Pres2Code   = AuxArray_Flt[NUC_AUX_PRES2CODE];
+
+   const int  NRho        = AuxArray_Int[NUC_AUX_NRHO  ];
+   const int  NEps        = AuxArray_Int[NUC_AUX_NEPS  ];
+   const int  NYe         = AuxArray_Int[NUC_AUX_NYE   ];
+   const int  NMode       = AuxArray_Int[NUC_AUX_NMODE ];
+
+   int  Mode      = NUC_MODE_ENGY;
+   real Dens_CGS  = Dens_Code * Dens2CGS;
+   real sEint_CGS = ( Eint_Code / Dens_Code ) * sEint2CGS - EnergyShift;
```

Comment:
Why do we need to subtract `EnergyShift` here?

Reply:
Because we add `EnergyShift` in the initial gas energy (see https://github.com/hyschive/gamer-fork/blob/ccsn--nuclear-eos/src/TestProblem/Hydro/Riemann/Init_TestProb_Hydro_Riemann.cpp#L421) to be consistent with FLASH. It helps to avoid negative internal energy. Given that the nuclear EoS driver requires the *real* gas energy, we must subtract this energy shift first.